### ISSUE TITLE
[LLD] LTO error in 17.0.0-rc1: "relocation refers to a symbol in a discarded section"

### ISSUE THREAD

**Issue by vient** [llvm/llvm-project#64429]
Hi, so I tried 17.0.0-rc1 on our codebase that has no problem building with 16.0.6 or GCC 12/13.
LTO build with clang and lld 17.0.0-rc1 produces error
```
ld.lld-17: error: relocation refers to a symbol in a discarded section: std::stack<long, std::deque<long, std::allocator<long>>>::~stack()
>>> defined in /usr/lib/xxx.a(yyy.o)
>>> referenced by regex_compiler.tcc:276 (/usr/bin/../lib/gcc/x86_64-linux-gnu/12.1.0/../../../../include/c++/12.1.0/bits/regex_compiler.tcc:276)
>>>               lto.tmp:(std::__detail::_Compiler<std::__cxx11::regex_traits<char>>::_M_quantifier())
>>> referenced by regex_automaton.tcc:228 (/usr/bin/../lib/gcc/x86_64-linux-gnu/12.1.0/../../../../include/c++/12.1.0/bits/regex_automaton.tcc:228)
>>>               lto.tmp:(std::__detail::_StateSeq<std::__cxx11::regex_traits<char>>::_M_clone())
clang++-17: error: linker command failed with exit code 1 (use -v to see invocation)
```

Here it what I found:
* Same LTO build using clang 16 instead of 17 (so still using lld 17) ends successfully.
* Same LTO build with ld (using LLVM gold plugin) ends successfully — that's why I think that the problem is in LLD itself and not any shared part with gold plugin.
* Same build without LTO ends successfully.
* Deleting some seemingly random code block from our codebase makes the problem go away. This code block does not use std::stack at all.

I did not manage to create small reproducer yet and I can't share our whole codebase. Creating this issue in hopes that someone knows what may be going on or can give advice on identifying problem (producing debug output from lld for example).

**Comment by vient** [comment]
Ok, it's hard to bisect, first try on `git bisect` failed to compile object files, and I can't bisect just lld because it works fine with object files from clang16. Don't know how to proceed with this.

**Comment by vient** [comment]
Bisected to https://github.com/llvm/llvm-project/commit/456468a08aac46642d5bf05701517d2f6503d7b8

@teresajohnson do you have an idea what is wrong here?

**Comment by teresajohnson** [comment]
@MaskRay for thoughts since this is only happening with lld.

It looks like that error happens in lld when there is a reference to a non-prevailing COMDAT group.

My change would have resulted in non-prevailing copies of linkonce/weak ODR functions no longer being internalized when the definition is in a native object. Instead they should be marked available_externally as we normally do for non-prevailing copies of these types of functions when the prevailing copy is in IR. I'm not totally sure how my change would result in instances of the above. In fact, per D135427 it looks like any local linkage symbol in a non-prevailing copy is converted to available_externally anyway.

@vient is yyy.o a native object or is it an LTO bitcode object? Also, I assume this is a ThinLTO build since my change should only affect that. Can you share your compile options?

**Comment by vient** [comment]
Yes, this is a ThinLTO build and yyy.o is a native object (xxx.a as a whole is third-party library which we can't recompile).

Minified link command looks like this
```
/usr/bin/clang++-17 \
    -flto=thin \
    -fuse-ld=lld \
    --ld-path=/usr/bin/ld.lld-17 \
    tree/main.cpp.o \
    tree/main_launcher.cpp.o \
    -o main \
    < a ton of -L... and archives from our codebase/boost/abseil/... omitted > \
    /usr/lib/xxx.a \
    -ldl \
    -lpthread
```

**Comment by vient** [comment]
Gathered some info using this procedure https://stackoverflow.com/a/49626057
Does not seem useful but just in case
```
$ readelf -sW /usr/lib/xxx.a | grep stack | c++filt
   133: 0000000000000000     5 FUNC    WEAK   DEFAULT   49 std::stack<long, std::deque<long, std::allocator<long> > >::~stack()
   134: 0000000000000000     5 FUNC    WEAK   DEFAULT   49 std::stack<long, std::deque<long, std::allocator<long> > >::~stack()


$ readelf -SW /usr/lib/xxx.a | grep -A100 yyy.o
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
...
  [49] .text._ZNSt5stackIlSt5dequeIlSaIlEEED2Ev PROGBITS        0000000000000000 034ec0 000005 00 AXG  0   0 16


$ readelf -g /usr/lib/xxx.a | grep -A100 yyy.o
...
COMDAT group section [    5] `_ZNSt5stackIlSt5dequeIlSaIlEEED5Ev' [_ZNSt5stackIlSt5dequeIlSaIlEEED5Ev] contains 1 sections:
   [Index]    Name
   [   49]   .text._ZNSt5stackIlSt5dequeIlSaIlEEED2Ev


$ readelf -SW /usr/lib/xxx.a | grep -A100 yyy.o
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
...
  [ 5] _ZNSt5stackIlSt5dequeIlSaIlEEED5Ev GROUP           0000000000000000 000060 000008 04     124  80  4


$ readelf -sW /usr/lib/xxx.a | grep -A100 yyy.o | c++filt
Symbol table '.symtab' contains 310 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
...
    80: 0000000000000000     0 SECTION LOCAL  DEFAULT    5
```

**Comment by teresajohnson** [comment]
Thanks for the detail. So it does look like this symbol is defined in a COMDAT in a native object. Presumably lld decided that was the prevailing copy, which would have triggered the change from the patch in question. But I'm not sure why that would provoke the error here. It sounds from the error like the COMDAT in the native object was discarded, but not sure why that would be if it was prevailing.

Can you add -Wl,-y,_ZNSt5stackIlSt5dequeIlSaIlEEED2Ev and send the results (it should give a list of where that symbol is defined and referenced from).

**Comment by vient** [comment]
```
/usr/lib/xxx.a(yyy.o): lazy definition of _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
/usr/lib/xxx.a(yyy.o): definition of _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tree/base/libbase.a(regex.cpp.o): reference to _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
/usr/lib/xxx.a(yyy.o): reference to _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
/usr/lib/xxx.a(yyy.o): reference to _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
lto.tmp: reference to _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
```

**Comment by teresajohnson** [comment]
@MaskRay is probably the best person to help track this down. I believe he is out of office for a few days, hopefully he can help when he gets back later this week.

The definition in the native yyy.o object is the one that appears to be prevailing, which would explain why my patch changed the behavior - with my patch we would no longer keep another local copy of this symbol but rather would end up with a reference to the prevailing copy. However, it appears as though the COMDAT in yyy.o is subsequently discarded by lld, and that part I don't understand since it is prevailing.

**Comment by MaskRay** [comment]
> I did not manage to create small reproducer yet and I can't share our whole codebase. Creating this issue in hopes that someone knows what may be going on or can give advice on identifying problem (producing debug output from lld for example).

The issue has only mentioned the symptom but gives no reproduce instructions. Such problems are nearly impossible to troubleshoot for someone who cannot access your codebase. Is it feasible to relink the program with `LLD_REPRODUCE=/tmp/rep.tar` or `-Wl,--reproduce=/tmp/rep.tar` and upload the reproduce tarball somewhere?

**Comment by vient** [comment]
Did not occur to me at first that I can extract needed object file from xxx.a.
[Here is the repro](https://github.com/llvm/llvm-project/files/12361172/repr.zip), links fine with
```
/usr/bin/clang++-16 -flto=thin --ld-path=/usr/bin/ld.lld-16 main.cpp.o dxplib.o libregex_lib.a libstub_lib.a
```
but fails if you change lld version to 17.

**Comment by vient** [comment]
Does not reproduce if `dxplib.o libregex_lib.a` are swapped, even if you take full linker command line and place them last after all standard libraries - if dxplib.o is before libregex_lib.a, error will be produced.

**Comment by MaskRay** [comment]
tl;dr The input is invalid. There is nothing wrong on lld or LLVM LTO.

The symbol resolution rule hasn't changed between lld 16 and 17.
`error: relocation refers to a symbol in a discarded section` is due to (a) improved LTO and (b) invalid input file `tmp/repr/dxplib.o`.

`tmp/repr/dxplib.o` has invalid GRP_COMDAT sections groups (e.g. 3, 4, 5) that reference (`sh_info`) local symbols (78, 79, 80).
This is itself problematic (see https://maskray.me/blog/2021-07-25-comdat-and-section-group#grp_comdat).
What makes the input invalid is that the local symbols are STT_SECTION symbols with empty names (st_name==0).
Linkers simply use the names to eliminate duplicates. In this case, N-1 GRP_COMDAT section groups out of the N invalid groups are discarded

```
% readelf -p.comment tmp/repr/dxplib.o

String dump of section '.comment':
  [     1]  GCC: (GNU) 4.4.7 20120313 (Red Hat 4.4.7-3)
% readelf -WSsg tmp/repr/dxplib.o
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .group            GROUP           0000000000000000 000040 000008 04     124 125  4
  [ 2] .group            GROUP           0000000000000000 000048 000008 04     124 126  4
  [ 3] _ZNSt6vectorIlSaIlEED5Ev GROUP           0000000000000000 000050 000008 04     124  78  4
  [ 4] _ZNSt11_Deque_baseIlSaIlEED5Ev GROUP           0000000000000000 000058 000008 04     124  79  4
  [ 5] _ZNSt5stackIlSt5dequeIlSaIlEEED5Ev GROUP           0000000000000000 000060 000008 04     124  80  4
...
Symbol table '.symtab' contains 310 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name 
...
    78: 0000000000000000     0 SECTION LOCAL  DEFAULT    3 _ZNSt6vectorIlSaIlEED5Ev
    79: 0000000000000000     0 SECTION LOCAL  DEFAULT    4 _ZNSt11_Deque_baseIlSaIlEED5Ev
    80: 0000000000000000     0 SECTION LOCAL  DEFAULT    5 _ZNSt5stackIlSt5dequeIlSaIlEEED5Ev
```

(Note that section groups not named ".group" are abnormal.)

It turns out that LLVM 16 LTO does not discard `define linkonce_odr dso_local void @_ZNSt5stackIlSt5dequeIlSaIlEEED2Ev`  from bitcode files
```
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).0.preopt.bc: ---------------- W _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).1.promote.bc: ---------------- W _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).2.internalize.bc: ---------------- t _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).3.import.bc: ---------------- t _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).4.opt.bc: ---------------- t _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).5.precodegen.bc: ---------------- t _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
```

While LLVM 17 LTO gets improved and discards `define linkonce_odr dso_local void @_ZNSt5stackIlSt5dequeIlSaIlEEED2Ev`
```
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).0.preopt.bc: ---------------- W _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).1.promote.bc:                  U _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).2.internalize.bc:                  U _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).3.import.bc:                  U _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).4.opt.bc:                  U _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
tmp/repr/libregex_lib.a(lib.cpp.o at 24272).5.precodegen.bc:                  U _ZNSt5stackIlSt5dequeIlSaIlEEED2Ev
```

The reference will lead to `error: relocation refers to a symbol in a discarded section`.

If you cannot rebuild  `tmp/repr/dxplib.o`, a hack is to give these local symbols (used as group signatures) unique names.
Unfortunately this cannot be done with objcopy/llvm-objcopy.
